File: packages/sdk-js/src/resources/webhook.ts
/**
 * Webhook configuration resource — `get` / `update` / `test`.
 *
 * Manages the account's outbound-webhook configuration: the delivery URL, the
 * signing secret, and the auto-disable status. The config carries NO money, so
 * the `/v1` and `/v2` surfaces are byte-identical (no `meta.fx`); both are
 * exposed for API-surface symmetry (`client.webhook.*` defaults to `/v2`,
 * `client.v1.webhook.*` is the `/v1` alias).
 *
 * **Security note:** `get`/`update` return `webhook_secret` in CLEAR TEXT so you
 * can re-verify your local signature check — treat the response as sensitive and
 * keep it server-side. See {@link verifyWebhookSignature}.
 *
 * `update` and `test` are state-changing. They carry no idempotency key, so —
 * like the order mutations — they are issued with retry DISABLED (`retry: 0`): a
 * transient failure surfaces as a thrown error rather than a silent replay.
 */
import type { RequestFn } from "./orders.js";
import type { components } from "../types.gen.js";

/** The saved outbound-webhook configuration (secret returned in clear text). */
export type WebhookConfig = components["schemas"]["WebhookConfig"];
/** Body for `update` — provide at least one field; `""` clears a field. */
export type UpdateWebhookBody = components["schemas"]["UpdateWebhookBody"];
/** The result of `test()` — the HTTP status your endpoint returned. */
export type WebhookTestResult = components["schemas"]["WebhookTestResult"];

/** A webhook config surface bound to one API version prefix (`/v1` or `/v2`). */
class WebhookResource {
  constructor(
    private readonly request: RequestFn,
    private readonly prefix: "/v1" | "/v2",
  ) {}

  /** Fetch the current webhook configuration (includes the clear-text secret). */
  async get(): Promise<WebhookConfig> {
    const { data } = await this.request<WebhookConfig>(
      "GET",
      `${this.prefix}/webhook`,
    );
    return data as WebhookConfig;
  }

  /**
   * Set or clear the delivery URL and/or signing secret.
   *
   * Send at least one of `webhook_url` / `webhook_secret`; pass `""` to clear a
   * field. A successful update also clears any auto-disable state. Not
   * auto-retried (state-changing, no idempotency key).
   */
  async update(body: UpdateWebhookBody): Promise<WebhookConfig> {
    const { data } = await this.request<WebhookConfig>(
      "PATCH",
      `${this.prefix}/webhook`,
      { body, retry: 0 },
    );
    return data as WebhookConfig;
  }

  /**
   * Send a `webhook.test` event to your configured URL and report the HTTP
   * status your endpoint returned. Requires a configured URL and is heavily
   * rate-limited. Not auto-retried (state-changing, no idempotency key).
   */
  async test(): Promise<WebhookTestResult> {
    const { data } = await this.request<WebhookTestResult>(
      "POST",
      `${this.prefix}/webhook/test`,
      { retry: 0 },
    );
    return data as WebhookTestResult;
  }
}

/** The `/v2` webhook config surface (default; byte-identical to `/v1`). */
export class V2WebhookResource extends WebhookResource {
  constructor(request: RequestFn) {
    super(request, "/v2");
  }
}

/** The `/v1` webhook config surface. */
export class V1WebhookResource extends WebhookResource {
  constructor(request: RequestFn) {
    super(request, "/v1");
  }
}

File: packages/sdk-py/src/smscode/resources/webhook.py
from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Any, Literal

from smscode.errors import InvalidResponseError
from smscode.models import WebhookConfig, WebhookTestResult
from smscode.types import ApiResult

SyncRequest = Callable[..., ApiResult[Any]]
AsyncRequest = Callable[..., Awaitable[ApiResult[Any]]]
ApiPrefix = Literal["/v1", "/v2"]


def _update_body(
    *,
    webhook_url: str | None = None,
    webhook_secret: str | None = None,
) -> dict[str, str]:
    body: dict[str, str] = {}
    if webhook_url is not None:
        body["webhook_url"] = webhook_url
    if webhook_secret is not None:
        body["webhook_secret"] = webhook_secret
    return body


def _decode_webhook_config(data: object) -> WebhookConfig:
    if not isinstance(data, dict):
        raise InvalidResponseError("The webhook config response is malformed.")
    consecutive = data.get("webhook_consecutive_failures")
    return WebhookConfig(
        webhook_url=data.get("webhook_url") if isinstance(data.get("webhook_url"), str) else None,
        webhook_secret=(
            data.get("webhook_secret") if isinstance(data.get("webhook_secret"), str) else None
        ),
        webhook_disabled_at=(
            data.get("webhook_disabled_at")
            if isinstance(data.get("webhook_disabled_at"), str)
            else None
        ),
        webhook_disabled_reason=(
            data.get("webhook_disabled_reason")
            if isinstance(data.get("webhook_disabled_reason"), str)
            else None
        ),
        webhook_consecutive_failures=consecutive if type(consecutive) is int else 0,
        raw=dict(data),
    )


def _decode_webhook_test_result(data: object) -> WebhookTestResult:
    if not isinstance(data, dict) or type(data.get("status_code")) is not int:
        raise InvalidResponseError("The webhook test response is malformed.")
    return WebhookTestResult(status_code=data["status_code"], raw=dict(data))


class WebhookResource:
    def __init__(self, request: SyncRequest, prefix: ApiPrefix) -> None:
        self._request = request
        self._prefix = prefix

    def get(self) -> WebhookConfig:
        return _decode_webhook_config(self._request("GET", f"{self._prefix}/webhook").data)

    def update(
        self,
        *,
        webhook_url: str | None = None,
        webhook_secret: str | None = None,
    ) -> WebhookConfig:
        return _decode_webhook_config(
            self._request(
                "PATCH",
                f"{self._prefix}/webhook",
                json=_update_body(webhook_url=webhook_url, webhook_secret=webhook_secret),
                retry=0,
            ).data
        )

    def test(self) -> WebhookTestResult:
        return _decode_webhook_test_result(
            self._request("POST", f"{self._prefix}/webhook/test", retry=0).data
        )


class V2WebhookResource(WebhookResource):
    def __init__(self, request: SyncRequest) -> None:
        super().__init__(request, "/v2")


class V1WebhookResource(WebhookResource):
    def __init__(self, request: SyncRequest) -> None:
        super().__init__(request, "/v1")


class AsyncWebhookResource:
    def __init__(self, request: AsyncRequest, prefix: ApiPrefix) -> None:
        self._request = request
        self._prefix = prefix

    async def get(self) -> WebhookConfig:
        return _decode_webhook_config((await self._request("GET", f"{self._prefix}/webhook")).data)

    async def update(
        self,
        *,
        webhook_url: str | None = None,
        webhook_secret: str | None = None,
    ) -> WebhookConfig:
        return _decode_webhook_config(
            (
                await self._request(
                    "PATCH",
                    f"{self._prefix}/webhook",
                    json=_update_body(webhook_url=webhook_url, webhook_secret=webhook_secret),
                    retry=0,
                )
            ).data
        )

    async def test(self) -> WebhookTestResult:
        return _decode_webhook_test_result(
            (await self._request("POST", f"{self._prefix}/webhook/test", retry=0)).data
        )


class AsyncV2WebhookResource(AsyncWebhookResource):
    def __init__(self, request: AsyncRequest) -> None:
        super().__init__(request, "/v2")


class AsyncV1WebhookResource(AsyncWebhookResource):
    def __init__(self, request: AsyncRequest) -> None:
        super().__init__(request, "/v1")
